File: examples/LCQP/warm_up.cpp
// Copyright 2020 Jonas Hall

#include <iostream>
#include <qpOASES.hpp>
USING_NAMESPACE_QPOASES

int main() {
    std::cout << "Preparing warm up problem...\n";

    /* Setup data of first QP. */
    real_t H[2*2] = { 2.0, 0.0, 0.0, 2.0 };
    real_t g[2] = { -2.0, -2.0 };
    real_t lb[2] = { 0.0, 0.0 };
    real_t ub[2] = { 10000.0, 10000.0 };
    real_t S1[1*2] = {1.0, 0.0};
    real_t S2[1*2] = {0.0, 1.0};

    real_t A[1*2] = { 1.0, 0.0 };
    real_t lbA[1] = { -10.0 };
    real_t ubA[1] = { 100.0};

    // Initial guess
    real_t x0[2] = { 1.0 + EPS, 1.0 };

    int_t nV = 2;
    int_t nC = 1;
    int_t nComp = 1;

    LCQProblem lcqp( nV, nC, nComp );

	Options options;
    options.initialComplementarityPenalty = 1.5;
    options.complementarityPenaltyUpdate = 2;
	lcqp.setOptions( options );

	int_t nWSR = 10000000;

    // Solve first LCQP
	returnValue retVal = lcqp.solve( H, g, A, lb, ub, lbA, ubA, S1, S2, nWSR, 0, x0 );

    if (retVal != SUCCESSFUL_RETURN)
    {
        printf("Termination ended without success.\n");
        return 1;
    }

    // Get solutions
    real_t xOpt[2];
	real_t yOpt[2];
	lcqp.getPrimalSolution( xOpt );
	lcqp.getDualSolution( yOpt );
	printf( "\nxOpt = [ %e, %e ];  yOpt = [ %e, %e ];  objVal = %e\n\n",
			xOpt[0],xOpt[1],yOpt[0],yOpt[1],lcqp.getObjVal() );

    return 0;
}

File: interfaces/python/examples/warm_up.py
import lcqpow
import numpy as np


print("Preparing dense warm up problem...")

# Setup data of first QP.
H = np.array([[2.0, 0.0], [0.0, 2.0]])
g = np.array([-2.0, -2.0])
S1 = np.array([[1.0, 0.0]])
S2 = np.array([[0.0, 1.0]])

x0 = np.array([1.0, 1.0])
y0 = np.array([0.0, 0.0, 0.0, 0.0])

nV = 2 
nC = 0
nComp = 1
lcqp = lcqpow.LCQProblem(nV=nV, nC=nC, nComp=nComp)

options = lcqpow.Options()
options.setPrintLevel(lcqpow.PrintLevel.INNER_LOOP_ITERATES)
options.setQPSolver(lcqpow.QPSolver.QPOASES_DENSE)
lcqp.setOptions(options)

retVal = lcqp.loadLCQP(H=H, g=g, S1=S1, S2=S2, x0=x0, y0=y0)
if retVal is not lcqpow.ReturnValue.SUCCESSFUL_RETURN:
    print("Failed to load LCQP.")

retVal = lcqp.runSolver()

if retVal is not lcqpow.ReturnValue.SUCCESSFUL_RETURN:
    print("Failed to solve LCQP.")

xOpt = np.zeros(2)
yOpt = np.zeros(nV+nC+2*nComp) 
stats = lcqpow.OutputStatistics()
lcqp.getPrimalSolution(xOpt)
lcqp.getDualSolution(yOpt)
lcqp.getOutputStatistics(stats)
print("xOpt = ", xOpt)
print("yOpt = ", yOpt)
print("i = ", stats.getIterTotal())
print("k = ", stats.getIterOuter())
print("rho = ", stats.getRhoOpt())
print("WSR = ", stats.getSubproblemIter())